### PR TITLE
Improved retry and other changes around mongo 4.4 + fix testing

### PR DESCRIPTION
**Pull request by hpidcock**
This PR:
- Bumps to github.com/juju/mgo/v3
- Fixes all the unit tests (skips a few that don't work in mongo 4.4 or features we don't use i.e. MapReduce)
- Improve sstxn retrying on retry-able errors (added error code handling here, will update juju/txn library which has the actual retry logic)
- sstxn txn log writes need the same error handling as any other write op.
- Improve CommitTransaction and AbortTransaction by bringing across the double-tap from Mongo Driver
- Simplify possible Query errors returned to callers by converting net.OpError to io.EOF, this improves test stability too.
- Non primary write errors force a cluster sync and discard the primary socket for that session. If there is a current transaction, abort it.

### CODE REVIEW

File: sstxn/sstxn.go [hunk 1]
```diff
@@ -408,6 +410,11 @@ func (r *Runner) applyRemove(op txn.Op, revno *int64) error {
 	} else if IsWriteConflict(err) {
 		r.logger.Tracef("remove op: %#v write conflict: %v", op, err)
 		return errWriteConflict
+	} else if mgo.IsTxnAborted(err) {
+		// This has triggered a server-side abort, so make sure the user
+		// understands the txn is aborted
+		r.logger.Tracef("insert op: %#v aborted: %v", op, err)
```
Comment (manadart) [inline sstxn/sstxn.go:416]: I wonder if it's worth logging any errors regarding an aborted transaction at a higher level. Then we don't have to bring in a slew of noise to see them.

**Review by manadart — COMMENTED** [review]
Looks good. One thought.

File: cluster_test.go [hunk 2]
```diff
@@ -34,8 +34,8 @@ import (
 	"sync"
```
Comment (manadart) [inline cluster_test.go:34]: `%s/Fallover/Failover/g` in this file?

File: sstxn/sstxn.go (hunk 1 above)
Comment (hpidcock) [reply sstxn/sstxn.go:416]: I've found that the changes I have in jujutxn and here have reduced the amount of "state changing too quickly"

Comment (hpidcock) [reply sstxn/sstxn.go:416]: But I agree, state changing too quickly means nothing but 🤷‍♂️ txn failed.

File: session_test.go [hunk 3]
```diff
@@ -3537,6 +3559,7 @@ func (s *S) TestDistinct(c *C) {
 }
 
 func (s *S) TestMapReduce(c *C) {
+	c.Skip("fix me")
```
Comment (wallyworld) [inline session_test.go:3562]: maybe a slightly longer message here

File: sstxn/sstxn.go [hunk 4]
```diff
@@ -157,7 +157,7 @@ func (r *Runner) runTxn(ops []txn.Op, id bson.ObjectId) error {
 	defer func() {
 		if !completed {
 			abortErr := r.db.Session.AbortTransaction()
-			if abortErr != nil {
+			if abortErr != nil && abortErr.Error() != "no transaction in progress" {
```
Comment (wallyworld) [inline sstxn/sstxn.go:160]: is the string literal from upstream? is there an error code or const that can be used?

**Review by wallyworld — APPROVED** [review]
Looks ok but we'd want to run some intensive testing once pulled into juju

**Pull request merged**